examples/token/src/lib.rs

```diff
@@ -0,0 +1,324 @@
+//! absolutely boilerplate, and should eventually be superseded by clever sugar
+//! in the form of proc macro attributes.
+//!
+//! - State loading.
+//! - State mutations.
+//! - Method dispatch.
+//! - Deserialization of parameters.
+//! - Serialization of return data.
+//!
+//! ## Addressing
+//!
+//! For simplicity, this actor uses ActorIDs in the balances and allowances
+//! HAMT (map) keys. This poses two problems:
+//!
+//! 1. A sender cannot send tokens to an inexistent account actor.
+//! 2. This actor is not reorg-safe (ActorIDs can change in the chain gets
+//!    reorg-ed).
+//!
+//! We _could_ do sophisticated things here to solve for these problems, by
+//! associating balances to class-{1,2,3} addresses only, and rejecting calls
+//! with class-0 (ID addresses).
+//!
+//! However, all of this will change when we introduce [class-4 addresses](https://github.com/filecoin-project/fvm-specs/blob/main/04-evm-mapping.md#proposed-solution-universal-stable-addresses)
+//! (reorg stable universal addresses), so it's not worth the complexity now,
+//! as this actor only exists for illustration purposes.
+//!
+//! ## NOT MEANT FOR PRODUCTION USAGE
+//!
+//! This actor should NEVER be deployed on the Filecoin network. It is purely
+//! meant for illustration and test purposes.
+//!
+
+use cid::multihash::{Code, MultihashDigest};
+use cid::Cid;
+use fvm_sdk as sdk;
+use fvm_sdk::blockstore::Blockstore;
+use fvm_sdk::message::{params_cbor, NO_DATA_BLOCK_ID};
+use fvm_shared::address::Address;
+use fvm_shared::bigint::bigint_ser::{BigIntDe, BigIntSer};
+use fvm_shared::bigint::{bigint_ser, Zero};
+use fvm_shared::blockstore::CborStore;
+use fvm_shared::econ::TokenAmount;
+use fvm_shared::encoding::tuple::*;
+use fvm_shared::encoding::{Cbor, RawBytes, DAG_CBOR};
+use fvm_shared::ActorID;
+use ipld_hamt::Hamt;
+
+/// A macro to abort execution by signalling a non-zero exit code.
+macro_rules! abort {
+    ($code:ident, $msg:literal $(, $ex:expr)*) => {
+        fvm_sdk::vm::abort(
+            fvm_shared::error::ExitCode::$code as u32,
+            Some(format!($msg, $($ex,)*).as_str()),
+        )
+    };
+}
+
+/// A macro to CBOR-encode a return value.
+macro_rules! encode {
+    ($ex:expr) => {
+        match RawBytes::serialize($ex) {
+            Ok(ret) => ret,
+            Err(err) => {
+                abort!(ErrSerialization, "failed to encode return value: {:?}", err)
+            }
+        }
+    };
+}
+
+/// The state object of the token actor.
+#[derive(Serialize_tuple, Deserialize_tuple, Clone, Debug)]
+pub struct State {
+    pub name: String,
+    pub symbol: String,
+    #[serde(with = "bigint_ser")]
+    pub max_supply: TokenAmount,
+    // map[ActorID] => TokenAmount
+    pub balances: Cid,
+    // TODO should never use ActorID as allowees, since it's an unstable address
+    //  make sure to check for this
+    //  map[ActorID] => map[Address]TokenAmount
+    // pub allowances: Cid,
```

Comment:
what if a failed resolve creates an ID and returns, and you just resolve in userspace before the send?

Reply:
ID addresses are not the solution because they are reorg-unstable. Because ID assignments can be reorged away, it is not safe to operate with a newly created ID address (e.g. an account actor, a miner, a payment channel) until finality (900 epochs). Otherwise, you could end up sending tokens to an address that gets reorg'ed away from you.´

The solution is to introduce class 4 addresses -- a universal, stable namespace for addresses: https://github.com/filecoin-project/fvm-specs/blob/main/04-evm-mapping.md#proposed-solution-universal-stable-addresses.

But we don't have those yet; that's why the token actor here uses the only other universal address option we have: the ID address.

Reply:
> ID addresses are not the solution because they are reorg-unstable. Because ID assignments can be reorged away, it is not safe to operate with a newly created ID address (e.g. an account actor, a miner, a payment channel) until finality (900 epochs). Otherwise, you could end up sending tokens to an address that gets reorg'ed away from you.´

This isn't generally true. You don't want to use an ID address in a _message_, but _state_ isn't re-org stable anyways. I.e.:

1. Use some form of stable address in a message.
2. Resolve to an ID address on-chain.
3. Use the ID address in on-chain datastructures.